### PR TITLE
Change reporting to TLS

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
Makes reporting connections over TLS.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by rainest** [comment]
I am doing something that breaks the test TLS connection. Possibly wrapping its accept in a goroutine is freezing it somehow? Not sure why. It receives a TLS client hello, but never responds with its own server hello. The struct indicates that it has not completed any handshakes:

[timeout.pcap.gz](https://github.com/Kong/kubernetes-ingress-controller/files/7695912/timeout.pcap.gz)
[conn.txt](https://github.com/Kong/kubernetes-ingress-controller/files/7695930/conn.txt)

**Comment by rainest** [comment]
Deadlock resolved by manually handshaking. I'm not particularly fond of the connection re-establishment behavior, but not immediately seeing how we can handle it better since we don't get an affirmative "yes, the connection is closed" `net.ErrClosed`, and what we do get isn't exported and would require us to scrape the syscall text. Is there some better way to determine that upstream closed the connection?

```
> github.com/kong/kubernetes-ingress-controller/v2/internal/util.Reporter.send() ./reports.go:105 (hits goroutine(12):3 total:3) (PC: 0x1a449c8)
	err: error(*crypto/tls.permanentError) *{
		err: net.Error(*net.OpError) *{
			Op: "write",
			Net: "tcp",
			Source: net.Addr(*net.TCPAddr) ...,
			Addr: net.Addr(*net.TCPAddr) ...,
			Err: error(*os.SyscallError) ...,},}
   100:	
   101:	func (r Reporter) send(signal string, uptime int) {
   102:		message := "<14>signal=" + signal + ";uptime=" +
   103:			strconv.Itoa(uptime) + ";" + r.serializedInfo
   104:		_, err := r.conn.Write([]byte(message))
=> 105:		if err != nil {
   106:			r.Logger.Errorf("failed to send report: %s", err)
   107:			r.conn = nil
   108:		}
   109:	}
(dlv) call err.Error()
> github.com/kong/kubernetes-ingress-controller/v2/internal/util.Reporter.send() ./reports.go:105 (PC: 0x1a449c8)
Values returned:
	~r0: "write tcp 127.0.0.1:46284->127.0.0.1:61833: write: broken pipe"

   100:	
   101:	func (r Reporter) send(signal string, uptime int) {
   102:		message := "<14>signal=" + signal + ";uptime=" +
   103:			strconv.Itoa(uptime) + ";" + r.serializedInfo
   104:		_, err := r.conn.Write([]byte(message))
=> 105:		if err != nil {
   106:			r.Logger.Errorf("failed to send report: %s", err)
   107:			r.conn = nil
   108:		}
   109:	}
```

**Comment by rainest** [comment]
Realized we don't actually need persistent connections and interruption handling: while the test interval is 1 second, and the actual interval does include `time.Second`, the actual interval is `3600 * time.Second`, not just `time.Second`.

Dunno if Splunk will even let us keep the idle connection for an hour, but there's no real need to, since reopening it once each hour doesn't raise the same performance issues as doing it once each second, and drastically simplifies testing.

This is ready for review, but we want to conduct a practical test where we run an image built from this branch against the actual reporting server to confirm we receive stats via the new listen. I need to get permissions to access that.

File: internal/util/reports.go
Comment (shaneutt) [inline internal/util/reports.go:16]: While we probably presume the endpoint will be negotiating `TLS1.3` by default and we may also assume that would never change in a negative direction, it seems we should simply require `tls.VersionTLS13` here anyway, no?
```suggestion
	tlsConf      = tls.Config{MinVersion: tls.VersionTLS13}
```

File: internal/util/reports.go
Comment (shaneutt) [inline internal/util/reports.go:17]: non-blocker: Before by nature of UDP we didn't really have any concerns about a timeout, but now it seems we do. Not a blocker, and I assume that the number we have here stems from testing we've done, but I just want to ensure that we're being deliberate about this setting as there are some possible implications.

File: internal/util/reports.go
Comment (shaneutt) [inline internal/util/reports.go:93]: From what I can see we're not managing R/W deadlines for this connection, so in theory this goroutine could reach an indefinite hang here under bad networking conditions, or during outage scenarios.

We should configure the write and read deadlines for this connection to avoid that situation, and then add logic to act on the error condition (ideally, by retrying).

Also, we should check the return value of the write to ensure that the number of bytes written matches the length of `message`.

File: internal/util/reports_test.go
Comment (shaneutt) [inline internal/util/reports_test.go:22]: I think a rename of this variable might be helpful and informative to new code readers at a glance:

```suggestion
	tlsCertForTestingReports = TLSPair{
```

I'm OK if you want to change that up, that just reads really well to me and I know exact what this cert is for at a glance.

File: internal/util/reports_test.go
Comment (shaneutt) [inline internal/util/reports_test.go:258]: no-blocker: no need to take action _right now_, but maybe a follow up. Helper functions like this might be good/useful in KTF (e.g. new package `pkg/utils/networking/tls`).

**Review by shaneutt — APPROVED** [review]
Looks good, only thing before merging is: let's make sure we take care of deadlines and write verifications for the tls connection.

File: internal/util/reports.go
Comment (rainest) [reply internal/util/reports.go:17]: The timeout choice was "arbitrary not infinite value", and 30s felt longer than I'd reasonably expect a TCP connection to complete its handshake. Are there some other criteria we should consider?

File: internal/util/reports.go
Comment (rainest) [reply internal/util/reports.go:16]: Dunno. I defaulted to the oldest version without significant currently known flaws because it is a minimum and the client will still use 1.3 opportunistically if available.

The upstream does support it, so the only issue I can think of are if there are instances behind intercept proxies that haven't implemented 1.3 yet. Asked to see if we have version data from the Kong reporters anywhere, but I doubt it. Absent that we can probably use 1.3 and revert if we see an unusual drop in reporting instances.

File: internal/util/reports.go
Comment (rainest) [reply internal/util/reports.go:99]: Deadline but no retry or length check: if send fails it just gives up. These are low priority and essentially retry by virtue of running every hour anyway: if one fails, wait around for the next ticker to try again.

File: internal/util/reports.go
Comment (shaneutt) [reply internal/util/reports.go:17]: The average server response times should be considered, and then it's _slightly_ stretching the scope of this PR but solid retry management and logging logic for this code might improve loose ends.

Comment (shaneutt) [reply internal/util/reports.go:17]: Possibly the average response time of the server, but I have every reason to believe 30s will cover that, especially with retries. nm

File: internal/util/reports.go
Comment (shaneutt) [reply internal/util/reports.go:99]: The deadline takes care of the most important thing :+1:

File: internal/util/reports.go
Comment (shaneutt) [inline internal/util/reports.go:99]: :+1:

**Review by shaneutt — APPROVED** [review]

**Comment by rainest** [comment]
https://github.com/Kong/kubernetes-ingress-controller/issues/2104 seems quite determined to fail here. Wait for that fix, merge main, and this should pass.

**Comment by shaneutt** [comment]
actually the most recent failure seemed unrelated:

```
        	            	Internal error occurred: failed calling webhook "validations.kong.konghq.com": Post "https://webhook-kong-validations-gateway.kong-system.svc:443/?timeout=10s": dial tcp 10.96.85.183:443: connect: connection refused
```

**Comment by rainest** [comment]
The heck, auto-merge merged it despite the failure?

Webhook thing looks like we probably need a connection test eventually on the webhook service before trying actually create resources.

**Pull request merged**
